### PR TITLE
APP-1164 update deploy script

### PR DESCRIPTION
**Pull request by Rekard0**
## Description

- Updated deploy script.
- Reorganized utils functions.

Task: [APP-1164](https://aragonassociation.atlassian.net/browse/APP-1164)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [x] I ran all tests with success and extended them if necessary.
- [x] I have updated the ``CHANGELOG.md`` file in the root folder.
- [x] I have tested my code on the test network.

### CODE REVIEW

File: packages/contracts/deploy/verify.ts [hunk 1]
```diff
@@ -3,71 +3,152 @@ import {DeployFunction} from 'hardhat-deploy/types';
...
   return new Promise(resolve => setTimeout(resolve, ms));
 }
 
 const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
+  console.log('\nVerifying contracts');
+
   const {deployments, ethers, run} = hre;
 
-  console.log('Verifying registry and factories contracts');
+  const msDelay = 2; // 3 minutes - Etherscan needs some time to process before trying to verify.
```
Comment (mathewmeconry) [inline packages/contracts/deploy/verify.ts:17]: rename to only `delay` or `minutesDelay`. the value in this constant isn't in ms

File: packages/contracts/deploy/verify.ts [hunk 2]
```diff
@@ -3,71 +3,152 @@ import {DeployFunction} from 'hardhat-deploy/types';
...
 
-  await run(TASK_ETHERSCAN_VERIFY, {
-    license: 'GPL-3.0',
-    solcInput: true,
-  });
+  await delay(msDelay);
 
-  const RegistryContract = await ethers.getContractAt(
+  // Prepare contracts and addresses
+  const managingDAOAddress = await getContractAddress('DAO', hre);
```
Comment (mathewmeconry) [inline packages/contracts/deploy/verify.ts:28]: wouldn't it be possible to get all the addresses from all contracts from the deployments folder? this way we cannot forget to add a contract to this verify list

**Review by mathewmeconry — APPROVED** [review]
LGTM except the 2 comments

File: packages/contracts/deploy/verify.ts (hunk 1 above)
Comment (Rekard0) [reply packages/contracts/deploy/verify.ts:17]: amended

**Review by novaknole — COMMENTED** [review]
First Round

File: packages/contracts/deploy/10_framework/00_ens.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  const daoNode = ethers.utils.namehash(daoDomain);
+
+  const officialEnsRegistryAddress = ENS_ADDRESSES[network.name];
+  let ensRegistryAddress;
+
+  if (!officialEnsRegistryAddress) {
+    const ens = await setupENS(deployer, daoDomain);
+    ensRegistryAddress = ens.address;
+  } else {
+    ensRegistryAddress = officialEnsRegistryAddress;
```
[the remark this answers is not in the document]
Comment (Rekard0) [reply packages/contracts/deploy/10_framework/00_ens.ts:42]: amended

File: packages/contracts/deploy/10_framework/00_ens.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+  const ensRegistryContract = await ethers.getContractAt(
+    'ENSRegistry',
+    ensRegistryAddress
+  );
+
+  // Approving future `ENSSubdomainRegistrar` address
+  // by using index 2, because the next deploy will be the logic address.
+  const futureAddress = await detemineDeployerNextAddress(2, deployer);
```
[the remark this answers is not in the document]
Comment (Rekard0) [reply packages/contracts/deploy/10_framework/00_ens.ts:52]: The deploy function is deploy the proxy and initializes it,
and the ens registrar check if it is approved, that is why we pre-approve

but yes, we can do it later, but then we need to deploy the contract manually in multiple step, if u we think we should go this route we can, if u think there is a big problem with predicting next address

[the remark this answers is not in the document]
Comment (novaknole) [reply packages/contracts/deploy/10_framework/00_ens.ts:52]: I get the point. have you looked into `DeterministicDeployment` of hardhat-deploy ? I don't know the risks yet though.

[the remark this answers is not in the document]
Comment (novaknole) [reply packages/contracts/deploy/10_framework/00_ens.ts:52]: In your case, there's a problem because it could happen that while you're deploying these and predicting the way you do, the same deployer outside just creates a tx. then your nonce (the time you get it with `getTransactionCount()` and while you actually call it might be mismatched). Hope that's clear.

File: packages/contracts/deploy/10_framework/10_dao-registry.ts [hunk 5]
```diff
@@ -1,16 +1,19 @@
...
+
+  // Get managing DAO address.
   const managingDAOAddress = await getContractAddress('DAO', hre);
+
+  // Get DAO's `ENSSubdomainRegistrar` contract.
   const ensSubdomainRegistrarAddress = await getContractAddress(
-    'ENSSubdomainRegistrar',
+    'DAO_ENSSubdomainRegistrar',
     hre
   );
```
[the remark this answers is not in the document]
Comment (Rekard0) [reply packages/contracts/deploy/10_framework/10_dao-registry.ts:18]: while for sure we can use the tags,
but it was hard to follow, while debugging
so we switched to numbering the files, that seemed more straight forward, 
WDYT?

[the remark this answers is not in the document]
Comment (novaknole) [reply packages/contracts/deploy/10_framework/10_dao-registry.ts:18]: Yes, but if somebody comes across this and doesn't know it's using filenames as the sort mechanism, hell knows what this causes

File: packages/contracts/deploy/10_framework/00_ens.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (mathewmeconry) [reply packages/contracts/deploy/10_framework/00_ens.ts:58]: fair point with the nonce and the deterministic address doesn't match anymore. But for now we only deploy to goerli and can restart without an issue. So I suggest we ignore this for now. 

I also guess we won't use github actions to deploy to mainnet anyway

File: packages/contracts/deploy/10_framework/10_dao-registry.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (mathewmeconry) [reply packages/contracts/deploy/10_framework/10_dao-registry.ts:18]: I agree with @Rekard0 to continue using numbering files instead of tags. Handling tags is a pain in the ass and awful for big deployments as ours

File: packages/contracts/deploy/10_framework/00_ens.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Rekard0) [reply packages/contracts/deploy/10_framework/00_ens.ts:58]: a task is created

[the remark this answers is not in the document]
Comment (Michael-A-Heuer) [reply packages/contracts/deploy/10_framework/00_ens.ts:58]: > a task is created

Please add to the task that we should consider removing the `_ens.isApprovedForAll(nodeOwner, address(this))` from the `initialize` function in `ENSSubdomainRegistrar`.
